DIMS/GenerateExcel.R

```diff
@@ -159,33 +185,58 @@ if (z_score == 1) {
   colnames(outlist)[startcol:ncol(outlist)] <- colnames_z
 
   # calculate robust scaler (Zscores minus outliers in Controls)
-  outlist_noZ$avg.ctrls <- 0
-  outlist_noZ$sd.ctrls  <- 0
+  outlist_robustZ$avg.ctrls <- 0
+  outlist_robustZ$sd.ctrls  <- 0
+  outlist_nooutliers$avg.ctrls <- 0
+  outlist_nooutliers$sd.ctrls <- 0
+  outlist_nooutliers$nr.ctrls <- 0
 
   # only calculate robust Z-scores if there are enough Controls
   if (length(control_col_ids) > 10) {
     for (metabolite_index in 1:nrow(outlist)) {
-      outlist_noZ$avg.ctrls[metabolite_index] <- mean(robust_scaler(outlist_noZ[metabolite_index, control_col_ids],
+      outlist_robustZ$avg.ctrls[metabolite_index] <- mean(robust_scaler(outlist_robustZ[metabolite_index, control_col_ids],
                                                                     control_col_ids, perc))
-      outlist_noZ$sd.ctrls[metabolite_index]  <-   sd(robust_scaler(outlist_noZ[metabolite_index, control_col_ids],
+      outlist_robustZ$sd.ctrls[metabolite_index]  <-   sd(robust_scaler(outlist_robustZ[metabolite_index, control_col_ids],
                                                                     control_col_ids, perc))
     }
   }
 
   # Make and add columns with robust zscores
   cnames_robust <- gsub("_Zscore", "_RobustZscore", colnames_z)
   for (i in intensity_col_ids) {
-    zscores_1col <- (as.numeric(as.vector(unlist(outlist_noZ[, i]))) - outlist_noZ$avg.ctrls) / outlist_noZ$sd.ctrls
-    outlist_noZ <- cbind(outlist_noZ, zscores_1col)
+    zscores_1col <- (as.numeric(as.vector(unlist(outlist_robustZ[, i]))) - outlist_robustZ$avg.ctrls) / outlist_robustZ$sd.ctrls
+    outlist_robustZ <- cbind(outlist_robustZ, zscores_1col)
   }
-  colnames(outlist_noZ)[startcol:ncol(outlist_noZ)] <- cnames_robust
+  colnames(outlist_robustZ)[startcol:ncol(outlist_robustZ)] <- cnames_robust
+
+  # calculate Z-scores after removal of outliers in Control samples
+  if (length(control_col_ids) > 10) {
+    for (metabolite_index in 1:nrow(outlist_nooutliers)) {
+      intensities_without_outliers <- remove_outliers_for_Zscore(as.numeric(outlist_nooutliers[metabolite_index, control_col_ids]), outlier_threshold)
+      outlist_nooutliers$avg.ctrls[metabolite_index] <-   mean(intensities_without_outliers)
+      outlist_nooutliers$sd.ctrls[metabolite_index]  <-     sd(intensities_without_outliers)
+      outlist_nooutliers$nr.ctrls[metabolite_index]  <- length(intensities_without_outliers)
+    }
+  }
+ 
+  cnames_nooutliers <- gsub("_Zscore", "_OutlierRemovedZscore", colnames_z)
+  for (i in intensity_col_ids) {
+    zscores_1col <- (as.numeric(as.vector(unlist(outlist_nooutliers[, i]))) - outlist_nooutliers$avg.ctrls) / outlist_nooutliers$sd.ctrls
+    outlist_nooutliers <- cbind(outlist_nooutliers, zscores_1col)
+  }
```

Comment:
It is often best to avoid for loops in R (very inefficient). 
```suggestion
outlist_nooutliers %>% apply(outlist_nooutliers, 2, function(col){
    (as.numeric(as.vector(unlist(col))) - outlist_nooutliers$avg.ctrls) / outlist_nooutliers$sd.ctrls
}) %>% cbind(outlist_nooutliers)
```